Index: linux+rh+chaos/include/asm-x86_64/page.h
===================================================================
--- linux+rh+chaos.orig/include/asm-x86_64/page.h
+++ linux+rh+chaos/include/asm-x86_64/page.h
@@ -10,7 +10,7 @@
 #define PAGE_MASK	(~(PAGE_SIZE-1))
 #define PHYSICAL_PAGE_MASK	(~(PAGE_SIZE-1) & __PHYSICAL_MASK)
 
-#define THREAD_ORDER 1 
+#define THREAD_ORDER 2 
 #define THREAD_SIZE  (PAGE_SIZE << THREAD_ORDER)
 #define CURRENT_MASK (~(THREAD_SIZE-1))
 
