Index: linux+rh+chaos/kernel/sched.c
===================================================================
--- linux+rh+chaos.orig/kernel/sched.c
+++ linux+rh+chaos/kernel/sched.c
@@ -1038,6 +1038,7 @@ inline int task_curr(const struct task_s
 {
 	return cpu_curr(task_cpu(p)) == p;
 }
+EXPORT_SYMBOL(task_curr);
 
 /* Used instead of source_load when we know the type == 0 */
 unsigned long weighted_cpuload(const int cpu)
