diff --git a/include/linux/Makefile.am b/include/linux/Makefile.am
index 0ea6cfe..44ffb41 100644
--- a/include/linux/Makefile.am
+++ b/include/linux/Makefile.am
@@ -3,7 +3,8 @@ COMMON_H =
 KERNEL_H = \
 	$(top_srcdir)/include/linux/dcache_compat.h \
 	$(top_srcdir)/include/linux/xattr_compat.h \
-	$(top_srcdir)/include/linux/vfs_compat.h
+	$(top_srcdir)/include/linux/vfs_compat.h \
+	$(top_srcdir)/include/linux/blkdev_compat.h
 
 USER_H =
 
diff --git a/include/linux/Makefile.in b/include/linux/Makefile.in
index 91e0c66..532fd60 100644
--- a/include/linux/Makefile.in
+++ b/include/linux/Makefile.in
@@ -90,7 +90,8 @@ SOURCES =
 DIST_SOURCES =
 am__kernel_HEADERS_DIST = $(top_srcdir)/include/linux/dcache_compat.h \
 	$(top_srcdir)/include/linux/xattr_compat.h \
-	$(top_srcdir)/include/linux/vfs_compat.h
+	$(top_srcdir)/include/linux/vfs_compat.h \
+	$(top_srcdir)/include/linux/blkdev_compat.h
 am__vpath_adj_setup = srcdirstrip=`echo "$(srcdir)" | sed 's|.|.|g'`;
 am__vpath_adj = case $$p in \
     $(srcdir)/*) f=`echo "$$p" | sed "s|^$$srcdirstrip/||"`;; \
