Index: b/Makefile.in
===================================================================
--- a/Makefile.in
+++ b/Makefile.in
@@ -47,7 +47,7 @@
 	$(top_srcdir)/module/unicode/Makefile.in \
 	$(top_srcdir)/module/zcommon/Makefile.in \
 	$(top_srcdir)/module/zfs/Makefile.in \
-	$(top_srcdir)/module/zpios/Makefile.in AUTHORS COPYING \
+	$(top_srcdir)/module/zpios/Makefile.in AUTHORS \
 	ChangeLog config/config.guess config/config.sub config/depcomp \
 	config/install-sh config/ltmain.sh config/missing
 subdir = .
